### DISCUSSION TITLE
Resource Processing

### DISCUSSION THREAD

**Discussion by thebonbon** [EveronLife/EveronLife#106] [Ideas]
Discussion on Resource Processing https://github.com/EveronLife/EveronLife/issues/4

There are a couple options

- [ ] Serverside
- [ ] Clientside
- [ ] Direct
- [ ] ID based

### Serverside
A public inventory for input / output
![image](https://user-images.githubusercontent.com/45827986/172403909-224e64c5-4d08-4ee2-880c-73f49f257f23.png)

**Pro**: 
Easy to implement, no complex server / client logic.
**Con**:
The input and output would mix with multiple people trying to process at the same time. That would lead to conflict even for friendly farmers and in the end you dont know how much of the stuff in the output is yours.

### Clientside
A private inventory for input / output
**Pro**: 
No mixing of input and output inventories. Multiple people can process at the same time.
**Con**:
Hard to implement. (Even possible?)
If someone wants to rob you, you could just put everything in the input inventory so its 100% save.
### Direct
Item get processed (replaced) directly in the Vehicle / Player Inventory.
**Pro**: 
Kinda easy to implement, no complex server / client logic.
Other people can steal from vehicle.
**Con**:
Needs check for vehicles if player is owner.
Not really immersive.

### ID based
Based on your ID a inventory with the same ID spawns as input / output inventory. This inventory can only be opened via Action
For every player a custom inventory spawns.
**Pro**: 
No mixing of input and output inventories. Multiple people can process at the same time.
Other people can steal the items if they steal your ID and open the input/output linked to your ID.
**Con**:
Hard to implement. Complex logic

--
For now I think we should start with direct processing and try to implement the ID based one in the future.
Any ideas for other options?

**Comment by PackJC** [comment]
I like ID based would promote a better expierience in the long run. I'm not sure how an ID could be stolen, but maybe there are ways that could prevent theft? 

Then again, I like the idea of the direct processing for open farming and a chance of stealing someone's crop.

**Reply by thebonbon** [reply]
ID will be an Item in the players inventory, so it could be stolen.

**Comment by Tomboru** [comment]
So I already said it on Discord.
A Serverside inventory is okay. -> It can give a lot of RP and fun.
So this should be kept.

But I think when you are at a public processor it should be possible to 
pay to process it "private". -> This can be done by a "Direct system" as you called it or ID based. (It doesn't really matter I think)

One thing would be still important when you can buy the processing machine and place it (as example) in your own house.

**Reply by Arkensor** [reply]
> One thing would be still important when you can buy the processing machine and place it (as example) in your own house.

I mean this will be up to the server owners, but it would destroy any kind of balancing if you can not control where people place processing steps. So I doubt this will be viable.

**Comment by Arkensor** [comment]
While I like the idea of the physical processing of items, this is going to be a tricky situation. What we need to add without a doubt is a simple hold F to process action. Everything happens in the player's inventory. This kind of advanced processing could be part of the advanced farming system. 

I have thought about this for like an hour now, but can't find a good flexible solution. This is a very tough problem and is something server owners will likely be very opinionated about. So I am not sure to what extent we should even handle this in the framework.
